File: src/lib/api/notifications.ts
/**
 * Notification API functions
 */

import { api } from '../api';

// Types matching backend schemas
export interface Notification {
    id: number;
    type: 'achievement' | 'streak' | 'wordlist' | 'reminder';
    title: string;
    message: string;
    isRead: boolean;
    createdAt: string;
}

export interface NotificationListResponse {
    notifications: Notification[];
    unreadCount: number;
}

export interface UnreadCountResponse {
    unreadCount: number;
}

/**
 * Get user's notifications
 */
export async function getNotifications(limit = 50): Promise<NotificationListResponse> {
    return api.get<NotificationListResponse>(`/api/notifications?limit=${limit}`);
}

/**
 * Get unread notification count
 */
export async function getUnreadCount(): Promise<UnreadCountResponse> {
    return api.get<UnreadCountResponse>('/api/notifications/unread-count');
}

/**
 * Mark a notification as read
 */
export async function markAsRead(notificationId: number): Promise<{ success: boolean }> {
    return api.patch<{ success: boolean }>(`/api/notifications/${notificationId}/read`, {});
}

/**
 * Mark all notifications as read
 */
export async function markAllAsRead(): Promise<{ success: boolean; count: number }> {
    return api.post<{ success: boolean; count: number }>('/api/notifications/read-all', {});
}

/**
 * Delete a notification
 */
export async function deleteNotification(notificationId: number): Promise<{ success: boolean }> {
    return api.delete<{ success: boolean }>(`/api/notifications/${notificationId}`);
}

export default {
    getNotifications,
    getUnreadCount,
    markAsRead,
    markAllAsRead,
    deleteNotification,
};

File: backend/app/api/notifications.py
"""Notification API endpoints."""

from fastapi import APIRouter, HTTPException

from app.api.deps import CurrentUser, DbSession
from app.schemas.notification import (
    NotificationListResponse,
    NotificationResponse,
    UnreadCountResponse,
)
from app.services.notification_service import NotificationService

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


@router.get("", response_model=NotificationListResponse)
async def get_notifications(
    current_user: CurrentUser,
    db: DbSession,
    limit: int = 50,
) -> NotificationListResponse:
    """Get user's notifications."""
    service = NotificationService(db)
    notifications = await service.get_user_notifications(current_user.id, limit=limit)
    unread_count = await service.get_unread_count(current_user.id)

    return NotificationListResponse(
        notifications=[
            NotificationResponse.from_notification(n) for n in notifications
        ],
        unreadCount=unread_count,
    )


@router.get("/unread-count", response_model=UnreadCountResponse)
async def get_unread_count(
    current_user: CurrentUser,
    db: DbSession,
) -> UnreadCountResponse:
    """Get count of unread notifications."""
    service = NotificationService(db)
    count = await service.get_unread_count(current_user.id)
    return UnreadCountResponse(unreadCount=count)


@router.patch("/{notification_id}/read")
async def mark_as_read(
    notification_id: int,
    current_user: CurrentUser,
    db: DbSession,
) -> dict:
    """Mark a notification as read."""
    service = NotificationService(db)
    success = await service.mark_as_read(notification_id, current_user.id)

    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")

    return {"success": True}


@router.post("/read-all")
async def mark_all_as_read(
    current_user: CurrentUser,
    db: DbSession,
) -> dict:
    """Mark all notifications as read."""
    service = NotificationService(db)
    count = await service.mark_all_as_read(current_user.id)
    return {"success": True, "count": count}


@router.delete("/{notification_id}")
async def delete_notification(
    notification_id: int,
    current_user: CurrentUser,
    db: DbSession,
) -> dict:
    """Delete a notification."""
    service = NotificationService(db)
    success = await service.delete_notification(notification_id, current_user.id)

    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")

    return {"success": True}




